### DISCUSSION TITLE
Dex integration with Ory kratos for user management

### DISCUSSION THREAD

**Discussion by AnirudhPokala** [dexidp/dex#2730] [Q&A] [answered]
Hi Team
We are working on a project, for which we are using dex. Its working cool for AuthN federations. 
But since dex isn't going to support local user management, we searched few of platforms and found that ory kratos has good features. 
But there isn't any connector for ory kratos. 
Can anyone please guide us how to proceed further?

**Comment by AnirudhPokala** [comment]
@sagikazarmark any idea about this?

**Comment by nabokihms** [comment, marked as answer]
Hello, @AnirudhPokala.
I think you need to use a combination of Ory Kratos and [Ory Hydra](https://github.com/ory/hydra)  + OIDC connector for Dex.

**Comment by sagikazarmark** [comment]
@AnirudhPokala when using Kratos, you'd want to use Dex as an OIDC provider in Kratos, not the other way around.

**Comment by juliuskrah** [comment]
For anyone finding this later: I built two working demos of exactly this combination (Dex/Hydra + Kratos, with the app talking OIDC to the provider and Kratos remaining the sole identity/session store behind it) to compare their tradeoffs.

**Key finding:** Dex's `authproxy` connector *does* work for this — trusts pre-verified headers, bridged from Kratos's session cookie via Ory Oathkeeper — but it never implements `connector.RefreshConnector`, so it cannot issue an OAuth2 `refresh_token`. That's a hard architectural limit of the connector, not a config gap (confirmed by reading the source and by live testing). Session renewal has to fall back to silently replaying the Authorization Code flow instead of a real refresh grant.

Ory Hydra doesn't have that limitation — it's a full OAuth2/OIDC provider, so it issues real, rotating `refresh_token`s for `offline_access`. The cost is that Hydra has no user store of its own, so you write small `/login` and `/consent` bridge routes that check the Kratos session and answer Hydra's Admin API on its behalf.

Full write-up + both working Next.js apps + Docker Compose stacks: juliuskrah/experiments#1 (see `oidc/specs/COMPARISON.md` for the side-by-side analysis).

Hope this saves someone the trip through Dex's source to find the refresh-token gap.
